File: app/src/main/java/fr/clerc/myapplication/RecyclerViewAdapter.java
package fr.clerc.myapplication;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import fr.clerc.myapplication.model.Feature;


public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder> {
    private List<Feature> featureList;
    private OnItemClickListener listener;

    public interface OnItemClickListener {
        void onItemClick(View view, Feature feature);
    }

    public RecyclerViewAdapter(List<Feature> featureList, OnItemClickListener listener) {
        this.featureList = featureList;
        this.listener = listener;
    }

    @Override
    public int getItemCount() {
        return this.featureList.size();
    }

    public void addFeatureList(List<Feature> featureList) {
        this.featureList.addAll(featureList);
        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_feature, parent, false);
        return new RecyclerViewAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, final int position) {
        final Feature feature = featureList.get(position);

        holder.image.setImageResource(R.drawable.metro_logo);
        holder.featureName.setText(feature.getProperties().getName());
        holder.featureStreet.setText(feature.getProperties().getStreet());
        holder.featureAddress.setText(feature.getProperties().getPostalCode() + " - " + feature.getProperties().getCity());

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onItemClick(v, feature);
            }
        });
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        final View itemView;
        final TextView featureName;
        final TextView featureStreet;
        final TextView featureAddress;
        final ImageView image;

        ViewHolder(View view) {
            super(view);
            itemView = view;
            featureName = view.findViewById(R.id.feature_name);
            featureStreet = view.findViewById(R.id.feature_street);
            featureAddress = view.findViewById(R.id.feature_address);
            image = view.findViewById(R.id.feature_type);
        }
    }
}

File: app/src/main/java/fr/clerc/myapplication/kotlin/RecyclerViewAdapter.kt
package fr.clerc.myapplication.kotlin

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import fr.clerc.myapplication.R
import fr.clerc.myapplication.kotlin.model.Feature
import kotlinx.android.synthetic.main.item_feature.view.*

class RecyclerViewAdapter(private val itemList: ArrayList<Feature>, val listener: (Feature) -> Unit): RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder>() {

    override fun getItemCount(): Int {
        return itemList.size
    }

    fun addFeatureList(list: ArrayList<Feature>) {
        itemList.clear()
        itemList.addAll(list)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_feature, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val feature = itemList[position]

        holder.name.text = feature.properties.name
        holder.street.text = feature.properties.street
        holder.address.text = feature.properties.postalCode.toString() + " - " + feature.properties.city
        holder.image.setImageResource(R.drawable.metro_logo)

        holder.itemView.setOnClickListener {
            listener(feature)
        }
    }

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val name: TextView = view.feature_name
        val address: TextView = view.feature_address
        val street: TextView = view.feature_street
        val image: ImageView = view.feature_type
    }
}